### PR TITLE
ci: add codacy

### PR DESCRIPTION
**Pull request by jithumkunjuraman**
Add codacy coverage report

### CODE REVIEW

File: .github/workflows/codacy-analysis.yml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        uses: codacy/codacy-analysis-cli-action@97bf5df3c09e75f5bcd72695998f96ebd701846e
+        with:
+          project-token: ${{ secrets.CODACY_PROJECT_TOKEN }}
+
+      - name: Run Codacy Coverage Script
+        run: |
+          go test -coverprofile=unit.coverage.out ./...
+          bash <(curl -Ls https://coverage.codacy.com/get.sh) report \
+              --force-coverage-parser go -r unit.coverage.out \
+              --project-token ${{ secrets.CODACY_PROJECT_TOKEN }}
```
Comment (mKeRix) [inline .github/workflows/codacy-analysis.yml:30]: I would move this into the workflow that runs the tests already, I don't think we need to run tests twice

File: .github/workflows/codacy-analysis.yml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+      - name: Run Codacy Analysis
+        uses: codacy/codacy-analysis-cli-action@97bf5df3c09e75f5bcd72695998f96ebd701846e
+        with:
+          project-token: ${{ secrets.CODACY_PROJECT_TOKEN }}
+
+      - name: Run Codacy Coverage Script
+        run: |
+          go test -coverprofile=unit.coverage.out ./...
+          bash <(curl -Ls https://coverage.codacy.com/get.sh) report \
```
Comment (mKeRix) [inline .github/workflows/codacy-analysis.yml:28]: I think you can use their coverage reporter action as well: https://github.com/codacy/codacy-coverage-reporter-action

File: .github/workflows/codacy-analysis.yml [hunk 3]
```diff
@@ -0,0 +1,30 @@
+name: Codacy Analysis
+
+on:
```
Comment (mKeRix) [inline .github/workflows/codacy-analysis.yml:3]: I believe these jobs/steps need to be limited to only run on pull_request from this same repo, not forks, as PRs from forks wouldn't have access to the project token secret

File: .github/workflows/build-deploy.yml [hunk 4]
```diff
@@ -26,22 +26,37 @@ jobs:
   #         version: v1.60.3
   #         args: --timeout 600s --verbose ./...
 
-  test:
+  codacy-analysis:
+    if: github.repository == github.event.pull_request.head.repo.full_name
```
Comment (mKeRix) [inline .github/workflows/build-deploy.yml:30]: Can you move this condition to just the codacy steps (steps have ifs too)? It still makes sense to run tests for forks imo

File: .github/workflows/build-deploy.yml [hunk 5]
```diff
@@ -26,22 +26,37 @@ jobs:
   #         version: v1.60.3
   #         args: --timeout 600s --verbose ./...
 
-  test:
+  codacy-analysis:
```
Comment (mKeRix) [inline .github/workflows/build-deploy.yml:29]: I would keep the previous naming `test`, as the codacy part is just reporting on the tests that were done

File: Makefile [hunk 6]
```diff
@@ -67,7 +67,7 @@ vet: ## Run go vet against code.
 
 .PHONY: test
 test: manifests generate fmt vet envtest ## Run tests.
-	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) --bin-dir $(LOCALBIN) -p path)" go test $$(go list ./... | grep -v /e2e | grep -v /test/utils) -coverprofile cover.out
+	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) --bin-dir $(LOCALBIN) -p path)" go test $$(go list ./... | grep -v /e2e | grep -v /test/utils) -coverprofile cover.out ./...
```
Comment (mKeRix) [inline Makefile:70]: What is the purpose of this change, isn't the go list command supposed to get all the modules to test (minus the excluded ones)?

Comment (jithumkunjuraman) [reply Makefile:70]: This I was testing when i encoundered some errors. I will revert back

File: .github/workflows/build-deploy.yml (hunk 5 above)
Comment (jithumkunjuraman) [reply .github/workflows/build-deploy.yml]: done

File: Makefile (hunk 6 above)
Comment (jithumkunjuraman) [reply Makefile]: done

File: .github/workflows/build-deploy.yml (hunk 4 above)
Comment (jithumkunjuraman) [reply .github/workflows/build-deploy.yml]: Done

**Comment by mKeRix** [comment]
For the future: this repo is configured with the [Angular commit conventions](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#-commit-message-guidelines), I would recommend to make heavier use of them. A CI change like this doesn't need to trigger a new release, since nothing would change in the functionality 🙂
